### PR TITLE
@alloy => Auto-enroll ACBers

### PR DESCRIPTION
**Pull request by mzikherman**
Ok I think this might be mergable!

It can still be optimized to move the collector profile fetching to a middleware.

### CODE REVIEW

File: src/apps/loyalty/server/index.tsx [hunk 1]
```diff
@@ -72,53 +73,63 @@ app.get("/inquiries", (req, res) => {
...
-        entrypoint: "/bundles/inquiries.js",
-        bootstrapData: `var DATA = ${JSON.stringify(data)}; var USER_DATA = ${JSON.stringify(req.user.toJSON())};`,
-      }))
-    })
+      if (info.loyalty_applicant_at) {
+        return res.redirect(req.baseUrl + "/thank-you")
+      }
+
+      if (info.confirmed_buyer_at) {
+        markCollector(req.user.get("accessToken"))
```
Comment (mzikherman) [inline src/apps/loyalty/server/index.tsx]: For some reason this is causing the dreaded `Error: Can't set headers after they are sent.` and a node server crash!

Comment (mzikherman) [inline src/apps/loyalty/server/index.tsx]: (or rather the `res.redirect` in the callback most likely)

File: src/apps/loyalty/server/index.tsx [hunk 2]
```diff
@@ -72,53 +73,63 @@ app.get("/inquiries", (req, res) => {
...
-      const info = JSON.parse(body)
+  fetchCollectorProfile(req.user.get("accessToken"))
+    .then((info) => {
       let html
-
-      if (!info.loyalty_applicant_at) {
-        if (!info.confirmed_buyer_at) {
-          html = renderToString(<ThreewThankYou userName={req.user.attributes.name} />)
-        } else {
+      if (info.loyalty_applicant_at) {
```
Comment (mzikherman) [inline src/apps/loyalty/server/index.tsx]: tweaking the logic, there was a small bug here

File: src/apps/loyalty/server/enroll_loyalty_applicant.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import * as request from "request"
+
+interface CollectorProfileData {
+  loyalty_applicant_at: string
+  confirmed_buyer_at: string
+}
+
+export function fetchCollectorProfile(accessToken: string): Promise<CollectorProfileData> {
```
Comment (mzikherman) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: I'll be refactoring this to be a middleware since all the routes need it (in addition to the user). Will probably attach it to `req.profile`

File: package.json [hunk 4]
```diff
@@ -74,6 +74,7 @@
     "express-session": "^1.15.1",
     "isomorphic-relay": "^0.7.4",
     "lodash": "^4.17.4",
+    "node-fetch": "^1.6.3",
```
Comment (mzikherman) [inline package.json]: I just did `yarn add node-fetch` but there was no `yarn.lock` changes. Is that correct? I think I'm missing something in the flow of adding a new dep...

File: src/apps/loyalty/server/enroll_loyalty_applicant.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import fetch from "node-fetch"
+
+export function fetchCollectorProfile(accessToken: string): any {
```
Comment (mzikherman) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: this returns a `Body` from the fetch lib, so should be getting typed properly.

I wasn't able to figure out how to import/type that properly tho so went with `any`.

File: src/apps/loyalty/server/index.tsx [hunk 6]
```diff
@@ -34,7 +34,7 @@ app.use(session({
   cookie: {},
 }))
 app.use(artsyPassport(Object.assign({}, process.env, {
-  CurrentUser,
+  CurrentUser: Backbone.Model,
```
Comment (mzikherman) [inline src/apps/loyalty/server/index.tsx]: We just need this so `get` and `set` 'work' in the passport lib...I wonder if that can be relaxed? https://github.com/artsy/artsy-passport/issues/104

**Comment by mzikherman** [comment]
So yea, this is mergable even w/o moving the stuff to a middleware (https://github.com/artsy/reaction-force/issues/87).

Additionally, this can probably use some specs?!

File: package.json (hunk 4 above)
Comment (orta) [inline package.json]: interesting! Looks like it's already in at that same spec level, you just made it specific as a dep now:

```
node-fetch@^1.0.1, node-fetch@^1.6.3:
  version "1.6.3"
  resolved "https://registry.yarnpkg.com/node-fetch/-/node-fetch-1.6.3.tgz#dc234edd6489982d58e8f0db4f695029abcd8c04"
  dependencies:
    encoding "^0.1.11"
    is-stream "^1.0.1"
```

It comes from `graphql-config-parser`

File: src/apps/loyalty/server/enroll_loyalty_applicant.ts (hunk 5 above)
Comment (orta) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: that stuff is a bit of a nightmare, if you remove the `any` it'll get typed for you from type inference ( you can test by removing it, then holding `alt` and clicking on fetchCollectorProfile.

Comment (orta) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: Specifically the API typing for fetch, I've had a few problems with it in Danger and Peril - but letting it infer has worked fine.

File: src/apps/loyalty/server/index.tsx [hunk 7]
```diff
@@ -72,53 +72,65 @@ app.get("/inquiries", (req, res) => {
...
+        .then(({data, props}) => {
+          const html = renderToString(<IsomorphicRelay.Renderer {...props} />)
+          const styles = styleSheet.rules().map(rule => rule.cssText).join("\n")
+          res.send(renderPage({
+            styles,
+            html,
+            entrypoint: "/bundles/inquiries.js",
+            bootstrapData: `var DATA = ${JSON.stringify(data)}; var USER_DATA = ${JSON.stringify(req.user.toJSON())};`,
+          }))
+        })
```
Comment (orta) [inline src/apps/loyalty/server/index.tsx]: warning, incoming untested psuedo-code. _If_ you want to look at using `async` / `await` for this code, it would look something like this:

```ts
app.get("/inquiries", async (req, res) => {
  if (!req.user) {
    return res.redirect(req.baseUrl + "/login")
  }

    const token = req.user.get("accessToken")
  const results = fetchCollectorProfile(token)
  const info = await results.json()

  if (info.loyalty_applicant_at) {
    return res.redirect(req.baseUrl + "/thank-you")
  }

  if (info.confirmed_buyer_at) {
      try {
        const applicant = markCollectorAsLoyaltyApplicant(token)
        const profile = applicant.json()
        return res.redirect(req.baseUrl + "/thank-you")
      } catch(err)
        console.error(err)
      }
  }

  const {data, props} = await IsomorphicRelay.prepareData({
    Container: Inquiries,
    queryConfig: new CurrentUserRoute(),
  }, res.locals.networkLayer)

  const html = renderToString(<IsomorphicRelay.Renderer {...props} />)
  const styles = styleSheet.rules().map(rule => rule.cssText).join("\n")
  
  res.send(renderPage({
    styles,
    html,
    entrypoint: "/bundles/inquiries.js",
    bootstrapData: `var DATA = ${JSON.stringify(data)}; var USER_DATA = ${JSON.stringify(req.user.toJSON())};`,
  }))
})
```

File: src/apps/loyalty/server/enroll_loyalty_applicant.ts (hunk 5 above)
Comment (mzikherman) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: What do you mean exactly? I got VSCode setup (and I'm loving it), I did what you said:

> you can test by removing it, then holding alt and clicking on fetchCollectorProfile.

and it didn't seem to do anything?

Comment (mzikherman) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: Actually it returns a [Promise<Response>](https://github.com/bitinn/node-fetch#fetchurl-options).

However, when specifying that, I get an error when doing:

```
.then(body => body.json())
    .then(info => {
      if (info.loyalty_applicant_at) {
```

with the error `'Property 'loyalty_applicant_at' does not exist on type 'Response'.'`

I'm sort of confused, I guess something is getting confused after I call `json()` on the `Response` (implementing the `Body` interface)? It _does_ work with `any` (or leaving it out altogether).

File: src/apps/loyalty/server/index.tsx (hunk 7 above)
Comment (mzikherman) [inline src/apps/loyalty/server/index.tsx]: Got it, is that just a preferable style over the indents/`then`/`catch` stuff from Promises?

If it's a bit more style-ish, I'll do that in a subsequent refactor PR (if that's cool).

File: src/apps/loyalty/server/enroll_loyalty_applicant.ts (hunk 5 above)
Comment (orta) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: Ah! You need `yarn add @types/node-fetch` 👍

Comment (orta) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: By adding that, and removing the `any` - it will run the inference correctly:

screen shot 2017-04-10 at 20 15 04

(looks like I meant `cmd`, not `alt` above too)

Comment (orta) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: right now TS is saying "well, I know nothing about fetch, so we'll let it do whatever it wants"

Comment (mzikherman) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: Interesting, so making that return type leads to the compilation error above, after calling `json()` and continuing the chain.

File: src/apps/loyalty/server/index.tsx (hunk 7 above)
Comment (orta) [inline src/apps/loyalty/server/index.tsx]: Aye, it's just a style thing - not pressure to to this, would make a nice learning tool as a separate PR 👍

File: src/apps/loyalty/server/enroll_loyalty_applicant.ts (hunk 5 above)
Comment (alloy) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: Definitely need to add those typings. Wanna pair on getting the types for this function in order?

File: src/apps/loyalty/server/index.tsx (hunk 6 above)
Comment (alloy) [inline src/apps/loyalty/server/index.tsx]: Good call, that’d be ace 👍

File: src/apps/loyalty/server/index.tsx [hunk 8]
```diff
@@ -72,53 +72,65 @@ app.get("/inquiries", (req, res) => {
...
       } else {
         return res.redirect(req.baseUrl) // baseUrl already has "/loyalty" so no need to append it.
       }
 
       const styles = styleSheet.rules().map(rule => rule.cssText).join("\n")
       res.send(renderPage({ styles, html, entrypoint: "" }))
-    }
-  })
+    })
+    .catch(err => console.error(err))
```
Comment (alloy) [inline src/apps/loyalty/server/index.tsx]: We should probably send these somewhere at some point.

**Comment by alloy** [comment]
Yeah some tests would be great, otherwise looks good 👌

File: package.json [hunk 9]
```diff
@@ -28,6 +28,7 @@
     "@types/express": "^4.0.35",
     "@types/express-session": "^0.0.32",
     "@types/jest": "^18.1.1",
+    "@types/node-fetch": "^1.6.7",
```
Comment (alloy) [inline package.json]: It looks like this is pulling in typings for `node-fetch` 1.6.7, whereas we appear to be installing `node-fetch` 1.6.3 below.

File: src/apps/loyalty/server/enroll_loyalty_applicant.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  return fetch(`${ARTSY_URL}/${path}`, {
+    headers: {
+      "X-Access-Token": accessToken,
+    },
+    method: verb,
+  }).then(resp => resp.json<T>())
+}
+
+export function fetchCollectorProfile(accessToken: string): Promise<CollectorProfileReponse> {
+  return gravity<CollectorProfileReponse>(accessToken, "api/v1/me/collector_profile")
```
Comment (l2succes) [inline src/apps/loyalty/server/enroll_loyalty_applicant.ts]: typo, `CollectorProfileResponse`
